src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,581 @@
+ *
+ *      // POSIX version
+ *      int mkdir(const char *pathname, mode_t mode);
+ *
+ * For most of the calls, the boilerplate code of splitting a path into
+ * dirname, basename and calling dfs_lookup on dirname could be abstracted
+ * into a function.
+ *
+ * 1. It would be useful to wrap the dfs_ calls in a posix-equivalent manner,
+ *    including errno, etc.
+ *    Since most of these POSIX calls are #included from sys, we might want
+ *    to use the prefix dfs_sys_ for these wrappers.
+ *
+ * 2. It would be nice if the calls used a d_hash for dfs_lookup calls,
+ *    since the parent directory will likely appear many times.
+ *    Since we need the dfs_t and dfs_obj_t in nearly all cases,
+ *    perhaps we could put a dfs_t, dfs_obj_t, and d_hash in a struct
+ *    that we can pass to each function.
+ * */
+
+/* TODO evaluate includes */
+#include <libgen.h>
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/xattr.h>
+#include <linux/xattr.h>
+#include <daos/checksum.h>
+#include <daos/common.h>
+#include <daos/event.h>
+#include <daos/container.h>
+#include <daos/array.h>
+
+#include "daos.h"
+#include "daos_fs.h"
+
+/** struct holding attributes for the dfs_sys calls */
+typedef struct dfs_sys dfs_sys_t;
+struct dfs_sys {
+	dfs_t			*dfs;		/** mounted filesystem */
+	struct d_hash_table	*dfs_hash;	/** optional lookup hash */
+
+	/* TODO pass obj separately to only the functions that need it? */
+	dfs_obj_t		*obj;		/** the obj */
+};
+
+/** struct holding parsed dirname, name, and cached parent obj */
+typedef struct dfs_sys_path {
+	char		*dir_name;	/** dirname(path) */
+	char		*name;		/** basename(path) */
+	dfs_obj_t	*parent;	/** dir_name obj */
+} dfs_sys_path_t;
+
+
+/**
+ * Hash handle for each entry.
+ */
+struct dfs_sys_hash_hdl {
+	d_list_t	entry;
+	dfs_obj_t	*oh;
+	char		*name;
+	size_t		num_refs; /* TODO datatype? */
+};
+
+/**
+ * Allocate a new dfs_sys_hash_hdl.
+ * TODO maybe just do this inline since only called once
+ */
+static struct dfs_sys_hash_hdl*
+dfs_sys_hash_hdl_new(void)
+{
+	struct dfs_sys_hash_hdl	*hdl;
+
+	hdl = malloc(sizeof(struct dfs_sys_hash_hdl));
+	if (hdl == NULL)
+		return NULL;
+	hdl->oh = NULL;
+	hdl->name = NULL;
+	hdl->num_refs = 0;
+
+	return hdl;
+}
+
+/*
+ * Delete a dfs_sys_hash_hdl.
+ */
+static void
+dfs_sys_hash_hdl_delete(struct dfs_sys_hash_hdl *hdl)
+{
+	if (hdl != NULL) {
+		if (hdl->oh != NULL)
+			dfs_release(hdl->oh);
+		if (hdl->name != NULL)
+			free(hdl->name);
+		free(hdl);
+	}
+}
+
+/*
+ * Get a dfs_sys_hash_hdl from the d_list_t.
+ */
+static inline struct dfs_sys_hash_hdl*
+dfs_sys_hash_hdl_obj(d_list_t *rlink)
+{
+	return container_of(rlink, struct dfs_sys_hash_hdl, entry);
+}
+
+/**
+ * Compare hash entry key.
+ * Simple string comparison of name.
+ */
+static bool
+dfs_sys_hash_key_cmp(struct d_hash_table *table, d_list_t *rlink,
+		     const void *key, unsigned int ksize)
+{
+	struct dfs_sys_hash_hdl	*hdl = dfs_sys_hash_hdl_obj(rlink);
+
```

Comment:
Check the string len first, then then compare them if the length is the same.

Reply:
Ah, yes. Much better!